% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
%\VignetteIndexEntry{rnaSeqMap primer}
%\VignetteKeywords{}
%\VignetteDepends{}
%\VignettePackage{rnaSeqMap}
%documentclass[12pt, a4paper]{article}
\documentclass[12pt]{article}

\usepackage{amsmath,pstricks}
\usepackage{hyperref}
\usepackage[numbers]{natbib}
\usepackage{color}
\usepackage[scaled]{helvet}
\renewcommand*\familydefault{\sfdefault}

\definecolor{NoteGrey}{rgb}{0.96,0.97,0.98}

\textwidth=6.2in
\textheight=9.5in
@ \parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-1in

\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}
\newcommand{\code}[1]{{\texttt{#1}}}

\author{Anna Lesniewska, Michal Okoniewski}
\begin{document}
\title{\code{rnaSeqMap}: RNASeq analyses using xmap database back-end}

\maketitle

\fcolorbox{black}{NoteGrey} {
\begin{minipage}{13.5cm}
\begin{center}
\textbf{ Vignette for v.2.11.1 - no Xmap database needed}
\end{center}

\end{minipage}
}

\tableofcontents
\newpage

\section{Recent changes and updates}
28.06.2012 - switched to GappedAlignments in RS class and used them to get coverage in ND
           - corrected all the rle into Rle
04.10.2011 - added data modification generators : generatorAddSquare(), generatorAdd(),generatorMultiply(), generatorTrunc(),
generatorPeak(), generatorSynth()

local coverage normalizations: standarizationNormalize(), densityNormalize(), min\_maxNormalize()

local coverage difference measures: ks\_test(), diff\_area(), diff\_derivative\_area(),qq\_plot(), qq\_derivative\_plot(), pp\_plot(), 
pp\_derivative\_plot(), hump\_diff1(), hump\_diff2()   
  
14.05.2011 - added parseGff3() and bam2sig() - MO, AL

\section{Introduction}

\code{rnaSeqMap} is a "middleware" library for RNAseq secondary analyses. It constitutes an API for such operations as:

\begin{itemize}
\item{access to any the reads of the experiment in possibly fastest time, according to any chromosome coordinates}
\item{accessing sets of reads according to genomic annotation in Ensembl}
\item{calculation of coverage and number of reads and transformations of those values}
\item{creating input for significance analysis algorithms - from edgeR and DESeq}
\item{precisely finding significant and consistent regions of expression}
\item{splicing analyses}
\item{visualizations of genes and expression regions}
\end{itemize}

The library is independent from the sequencing technology and reads mappina software. It needs either 
reads described as genome coordinates in the extended xmap database, or can alternatively read data as 
big as they can fit in the operational memory. The use with modified xmap database is recommended, as it 
overcomes memory limitations - thus the library can be efficiently run on not very powerful machines. 

The internal features of \code{rnaSeqMap} distinctive for this piece of software are:

\begin{itemize}
\item{sequencing reads and annotations in one common database - extended XMAP \cite{Yates07}}
\item{algorithm for finding irreducible regions of genomic expression - according to Aumann and Lindell \cite{Lindell03}}
\item{nucleotide-level splicing analysis}
\item{connectors for further gene- and region-level expression processing to DESeq \cite{Anders10} and edgeR \cite{Robinson09}}
\item{the routines for coverage, splicing index and region mining algorithm have been implemented in C for speed}
\end{itemize}

%}




\begin{thebibliography}{99}
\bibitem{Lindell03}  Aumann Y and Lindell Y, A Statistical Theory for Quantitative Association Rules, Journal of Intelligent Information Systems, 2003
\bibitem{Yates07} Yates T, Okoniewski M and Miller C, X:Map: annotation and visualization of genome structure for Affymetrix exon array analysis, Nucleic Acids Research, 2007
\bibitem{Gard07} Gardina P et al, Alternative splicing and differential gene expression in colon cancer detected by a whole genome exon array, BMC Genomics, 2006
\bibitem{Okon08} Okoniewski M et al, An annotation infrastructure for the analysis and interpretation of Affymetrix exon array data, Genome Biology, 2007
\bibitem{Anders10} Anders S and Huber W, Differential expression analysis for sequence count data, Nature Preceedings, 2010
\bibitem{Robinson09} Robinson M et al, edgeR: a Bioconductor package for differential expression analysis of digital gene expression data, Bioinformatics, Bioinformatics, 2009
\end{thebibliography}
\end{document}
